OpenRA.Mods.Common/Traits/Mobile.cs

```diff
@@ -183,78 +179,60 @@ public int TileSetMovementHash(TileSet tileSet)
-
-					// Non-sharable unit can enter a cell with shareable units only if it can crush all of them.
-					if (needsCellExclusively)
-						return false;
-					var crushables = a.TraitsImplementing<ICrushable>();
-					if (!crushables.Any())
-						return false;
-					foreach (var crushable in crushables)
-						if (!crushable.CrushableBy(Crushes, self.Owner))
-							return false;
-				}
+				if (Collides(self, actor, ignoreActor, canIgnoreMovingAllies))
+					return false;
 			}
 
 			return true;
 		}
 
-		public bool CanEnterCell(World world, Actor self, CPos cell, out int movementCost, Actor ignoreActor = null, CellConditions check = CellConditions.All)
+		public bool CollidesWithOtherActorsInCell(World world, Actor self, CPos cell, Actor ignoreActor = null, CellConditions check = CellConditions.All)
 		{
-			if ((movementCost = MovementCostForCell(world, cell)) == int.MaxValue)
+			// If the actor can share his cell and the cell he intends to enter
+			// has any free space for him, then it can move into it.
+			// (or whether it's told we don't need to check for transient actor blockages)
+			if ((SharesCell && world.ActorMap.HasFreeSubCell(cell)) ||
+				!check.HasFlag(CellConditions.TransientActors))
 				return false;
 
-			if (SharesCell && world.ActorMap.HasFreeSubCell(cell))
-				return true;
+			// If the actor cannot enter outrightly, we must check if it can
+			// crush the units inside the cell if they are enemies. If they are
+			// allies, we must check if they follow our direction and can ignore them
+			var canIgnoreMovingAllies = !check.HasFlag(CellConditions.BlockedByMovers);
 
-			if (check.HasFlag(CellConditions.TransientActors))
+			foreach (var actor in world.ActorMap.GetActorsAt(cell))
 			{
-				var canIgnoreMovingAllies = self != null && !check.HasFlag(CellConditions.BlockedByMovers);
-				var needsCellExclusively = self == null || Crushes == null || !Crushes.Any();
-				foreach (var a in world.ActorMap.GetUnitsAt(cell))
-				{
-					if (a == ignoreActor)
-						continue;
-
-					// Neutral/enemy units are blockers. Allied units that are moving are not blockers.
-					if (canIgnoreMovingAllies && self.Owner.Stances[a.Owner] == Stance.Ally && IsMovingInMyDirection(self, a)) continue;
-
-					// Non-sharable unit can enter a cell with shareable units only if it can crush all of them.
-					if (needsCellExclusively)
-						return false;
-					var crushables = a.TraitsImplementing<ICrushable>();
-					if (!crushables.Any())
-						return false;
-					foreach (var crushable in crushables)
-						if (!crushable.CrushableBy(Crushes, self.Owner))
-							return false;
-				}
+				if (Collides(self, actor, ignoreActor, canIgnoreMovingAllies))
+					return true;
 			}
 
-			return true;
+			return false;
+		}
+
+		public bool CanEnterCell(World world, Actor self, CPos cell, out int movementCost, Actor ignoreActor = null, CellConditions check = CellConditions.All)
```

Comment:
Could we move this next to its cousin that lack the out parameter, and reimplement the cousin to call this and just drop the out parameter on the floor to avoid duplication.